docs/e2e_testing.md

```diff
@@ -0,0 +1,412 @@
+btc_aggregates_workflow = bf.Workflow(
+    workflow_id='btc_aggregates_df_bq',
+    schedule_interval='@monthly',
+    definition=[
+        CalculateBitcoinAggregatesJob()
+    ])
+```
+
+There are two facts that you should pay special attention to:
+
+* The `e2e` environment is the default one. If you import the workflow without explicitly setting the environment, it is configured using the default `e2e` environment.
+* The `e2e` configuration ensures that each execution of the workflow uses a fresh, unique BigQuery dataset (of course,
+only if you execute the workflow in a separate process).
+
+Now, take a look at the e2e test for that workflow, where we use these facts:
+
+```python
+import unittest
+from datetime import datetime, timedelta
+
+from bigflow.testing import SpawnIsolateMixin
+
+from btc_aggregates.btc_aggregates_df_bq import (
+    btc_aggregates_workflow,
+    dataset as btc_aggregates_dataset,
+    BTC_TRANSACTIONS_TABLE_NAME)
+
+DATE = slice(0, 10)
+NOW = "2020-12-01 00:00:00"
+NOW_DT = datetime.fromisoformat(NOW + "+00:00")
+NOW_MINUS_ONE_MONTH = (NOW_DT - timedelta(weeks=4)).isoformat()[DATE]
+NOW_PLUS_ONE_MONTH = (NOW_DT + timedelta(weeks=4)).isoformat()[DATE]
+
+
+class BitcoinAggregatesWorkflowTestCase(SpawnIsolateMixin, unittest.TestCase):
+    def setUp(self) -> None:
+        btc_aggregates_dataset.create_table(f'''
+        CREATE TABLE IF NOT EXISTS {BTC_TRANSACTIONS_TABLE_NAME} (
+            fee FLOAT64,
+            block_timestamp_month DATE)
+        ''').run()
+
+    def tearDown(self) -> None:
+        btc_aggregates_dataset.delete_dataset()
+
+    def test_should_calculate_aggregates(self):
+        # given
+        btc_aggregates_dataset.insert(BTC_TRANSACTIONS_TABLE_NAME, [
+            {'fee': 0.5, 'block_timestamp_month': NOW_MINUS_ONE_MONTH},
+            {'fee': 1.0, 'block_timestamp_month': NOW[DATE]},
+            {'fee': 2.0, 'block_timestamp_month': NOW[DATE]},
+            {'fee': 4.0, 'block_timestamp_month': NOW_PLUS_ONE_MONTH},
+        ], partitioned=False).run(NOW)
+
+        # when
+        btc_aggregates_workflow.run(NOW)
+
+        # then
+        result = btc_aggregates_dataset.collect_list('''
+        SELECT fee_sum, count
+        FROM `{btc_aggregates}`
+        WHERE _PARTITIONTIME = '{dt}'
+        ''', record_as_dict=True).run(NOW)
+        self.assertEqual(result, [{'fee_sum': 3.0, 'count': 2}])
+
+    def test_should_handle_empty_transactions_table(self):
+        # when
+        btc_aggregates_workflow.run(NOW)
+
+        # then
+        result = btc_aggregates_dataset.collect_list('''
+        SELECT fee_sum, count
+        FROM `{btc_aggregates}`
+        WHERE _PARTITIONTIME = '{dt}'
+        ''', record_as_dict=True).run(NOW)
+        self.assertEqual(result, [{'fee_sum': 0.0, 'count': 0}])
+
+
+if __name__ == '__main__':
+    unittest.main()
+```
+
+Each of the two tests has the following schema:
+
+1. Preparing fake bitcoin transactions table and inserting test records.
```

Comment:
nit: We can add comments to the source code. It might help reader to understand where actually (what line) we a doing some thing.

Reply:
done